xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -4846,17 +4845,14 @@ public XWikiAttachment removeAttachment(XWikiAttachment attachment)
     public XWikiAttachment removeAttachment(XWikiAttachment attachmentToRemove, boolean toRecycleBin)
     {
         List<XWikiAttachment> list = getAttachmentList();
-        for (int i = 0; i < list.size(); i++) {
-            XWikiAttachment attachment = list.get(i);
-            if (attachmentToRemove.getFilename().equals(attachment.getFilename())) {
-                list.remove(i);
-                this.attachmentsToRemove.add(new XWikiAttachmentToRemove(attachment, toRecycleBin));
-                setMetaDataDirty(true);
-                return attachment;
-            }
+        if (!list.contains(attachmentToRemove)) {
+            attachmentToRemove = null;
+        } else {
+            ((XWikiAttachmentList) (list)).remove(attachmentToRemove);
         }
-
-        return null;
+        this.attachmentsToRemove.add(new XWikiAttachmentToRemove(attachmentToRemove, toRecycleBin));
```

Comment:
All this should be in the else otherwise you will end up adding a null XWikiAttachment to the list of attachments to remove.